### PR TITLE
PUBDEV-7202 add function to upload binary model

### PR DESCRIPTION
**Pull request by koniecsveta**
https://0xdata.atlassian.net/projects/PUBDEV/issues/PUBDEV-7202

### CODE REVIEW

File: h2o-core/src/main/java/hex/Model.java [hunk 1]
```diff
@@ -2679,6 +2680,28 @@ private InteractionPair(int v1, int v2, String[] v1Enums, String[] v2Enums) {
...
 
+    /**
+     * Uploads a binary model from a given frame.
+     * Note: binary model has to be created by the same version of H2O, import of a model from a different version will fail
+     * @param destinationFrame key of the frame containing the binary representation of the model on a local filesystem, HDFS, S3...
+     * @return instance of an H2O Model
+     * @throws IOException when reading fails
+     */
+    public static <M extends Model<?, ?, ?>> M uploadBinaryModel(String destinationFrame) throws IOException {
+        InputStream is = null;
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/Model.java]: is aways null and can be removed, as well as the finally block

File: h2o-core/src/main/java/hex/Model.java [hunk 2]
```diff
@@ -2679,6 +2680,28 @@ private InteractionPair(int v1, int v2, String[] v1Enums, String[] v2Enums) {
...
+        Frame fr = DKV.getGet(destinationFrame);
+      try (VecDataInputStream vecDataInputStream = new VecDataInputStream(fr.vec("bytes"))) {
+          final AutoBuffer ab = new AutoBuffer(vecDataInputStream);
+          @SuppressWarnings("unchecked")
+          M model = (M) Keyed.readAll(ab);
+          ab.close();
+          vecDataInputStream.close();
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/Model.java]: this is done by the try-with-resources

File: h2o-core/src/main/java/hex/Model.java (hunk 1 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/hex/Model.java]: fixed

File: h2o-core/src/main/java/hex/Model.java (hunk 2 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/hex/Model.java]: fixed

File: h2o-core/src/main/java/hex/Model.java [hunk 3]
```diff
@@ -2679,6 +2680,24 @@ private InteractionPair(int v1, int v2, String[] v1Enums, String[] v2Enums) {
...
+    /**
+     * Uploads a binary model from a given frame.
+     * Note: binary model has to be created by the same version of H2O, import of a model from a different version will fail
+     * @param destinationFrame key of the frame containing the binary representation of the model on a local filesystem, HDFS, S3...
+     * @return instance of an H2O Model
+     * @throws IOException when reading fails
+     */
+    public static <M extends Model<?, ?, ?>> M uploadBinaryModel(String destinationFrame) throws IOException {
+      Frame fr = DKV.getGet(destinationFrame);
+      try (VecDataInputStream vecDataInputStream = new VecDataInputStream(fr.vec(0))) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java]: I think VecDataInputStream is an overkill in this case, not needed. Please use vec.openStream() instead

File: h2o-persist-hdfs/src/main/java/water/persist/HdfsVecDataInputStream.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.hadoop.fs.Seekable;
+import water.fvec.Vec;
+
+import java.io.EOFException;
+import java.io.IOException;
+
+/**
+ * Seekable and PositionedReadable implementation of InputStream backed by a Vec data source.
+ */
+public class HdfsVecDataInputStream extends VecDataInputStream implements Seekable, PositionedReadable {
```
Comment (michalkurka) [inline h2o-persist-hdfs/src/main/java/water/persist/HdfsVecDataInputStream.java]: This class shouldn't be renamed

File: h2o-core/src/main/java/hex/Model.java (hunk 3 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/hex/Model.java]: fixed

File: h2o-persist-hdfs/src/main/java/water/persist/HdfsVecDataInputStream.java (hunk 4 above)
Comment (koniecsveta) [reply h2o-persist-hdfs/src/main/java/water/persist/HdfsVecDataInputStream.java]: fixed

File: h2o-py/h2o/h2o.py [hunk 5]
```diff
@@ -1409,6 +1409,18 @@ def download_model(model, path=""):
     path = os.path.join(os.getcwd() if path == "" else path, model.model_id)
     return api("GET /3/Models.fetch.bin/%s" % model.model_id, save_to=path)
 
+def upload_model(path):
+    """
+    Upload a model from the provided local path to the H2O cluster.
```
Comment (michalkurka) [inline h2o-py/h2o/h2o.py]: please add a reference to the corresponding download/save function

maybe we should stress this is for binary models (not MOJO) and make it even clearer that local path means where the python client is running (I remember confusion about the difference of save/download model)

File: h2o-py/h2o/h2o.py [hunk 6]
```diff
@@ -1409,6 +1409,18 @@ def download_model(model, path=""):
...
+    
+    :param path: A path specifying the location of the model to upload
+    
+    :returns: a new :class:`H2OEstimator` object.
+    """
+    response = api("POST /3/PostFile.bin", filename=path)
```
Comment (michalkurka) [inline h2o-py/h2o/h2o.py]: perfect!

File: h2o-r/h2o-package/R/import.R [hunk 7]
```diff
@@ -351,6 +351,20 @@ h2o.loadModel <- function(path) {
   h2o.getModel(res$model_id$name)
 }
 
+h2o.upload_model <- function(path) {
```
Comment (michalkurka) [inline h2o-r/h2o-package/R/import.R]: please add documentation

**Comment by michalkurka** [comment]
@koniecsveta thank you for making the suggested changes, I believe the PR is good to be merged - only documentation should be added/improved

File: h2o-r/h2o-package/R/import.R (hunk 7 above)
Comment (koniecsveta) [reply h2o-r/h2o-package/R/import.R]: updated also save_model/saveModel/download_model documentation here, hence closing this one: https://github.com/h2oai/h2o-3/pull/4367/

**Comment by michalkurka** [comment]
Looks good! Thank you for making the changes, one last thing:

@koniecsveta 

CRAN check is failing on 
```
[2020-03-27T17:23:40.988Z] Undocumented arguments in documentation object 'h2o.upload_model'

[2020-03-27T17:23:40.988Z]   ‘path’

[2020-03-27T17:23:40.988Z] Documented arguments not in \usage in documentation object 'h2o.upload_model':

[2020-03-27T17:23:40.988Z]   ‘path:’

[2020-03-27T17:23:40.988Z] 
```

**Comment by koniecsveta** [comment]
> Looks good! Thank you for making the changes, one last thing:
> 
> @koniecsveta
> 
> CRAN check is failing on
> 
> ```
> [2020-03-27T17:23:40.988Z] Undocumented arguments in documentation object 'h2o.upload_model'
> 
> [2020-03-27T17:23:40.988Z]   ‘path’
> 
> [2020-03-27T17:23:40.988Z] Documented arguments not in \usage in documentation object 'h2o.upload_model':
> 
> [2020-03-27T17:23:40.988Z]   ‘path:’
> 
> [2020-03-27T17:23:40.988Z] 
> ```

it failed again but on something else:

18:30:21    URL 'https://CRAN.R-project.org/src/contrib/Meta/current.rds': status was 'Transferred a partial file'
18:30:21  Error in readRDS(con) : cannot read from connection
18:30:21  Execution halted

will reexecute it one more time

**Pull request merged**